xls/public/c_api_vast.cc

```diff
@@ -616,4 +626,105 @@ struct xls_vast_statement* xls_vast_statement_block_add_nonblocking_assignment(
   return reinterpret_cast<xls_vast_statement*>(cpp_assignment);
 }
 
+struct xls_vast_module_port** xls_vast_verilog_module_get_ports(
+    struct xls_vast_verilog_module* m, size_t* out_count) {
+  auto* cpp_module = reinterpret_cast<xls::verilog::Module*>(m);
+  absl::Span<const xls::verilog::ModulePort> ports = cpp_module->ports();
+  *out_count = ports.size();
+  if (ports.empty()) {
+    return nullptr;
+  }
+  // Allocate an array owned by the caller (to be freed by
+  // xls_vast_verilog_module_free_ports).
+  auto** result = new xls_vast_module_port*[ports.size()];
+  for (size_t i = 0; i < ports.size(); ++i) {
+    // Cast away const to satisfy the C API – callers must treat the objects as
+    // read-only.
+    result[i] = reinterpret_cast<xls_vast_module_port*>(
+        const_cast<xls::verilog::ModulePort*>(&ports[i]));
+  }
+  return result;
+}
+
+void xls_vast_verilog_module_free_ports(struct xls_vast_module_port** ports,
+                                        size_t /*count*/) {
+  // Only the outer array was allocated – the pointed-to ModulePort objects are
+  // owned by the parent Module.
+  delete[] ports;
+}
+
+xls_vast_module_port_direction xls_vast_verilog_module_port_get_direction(
+    struct xls_vast_module_port* port) {
+  auto* cpp_port = reinterpret_cast<xls::verilog::ModulePort*>(port);
+  switch (cpp_port->direction) {
+    case xls::verilog::ModulePortDirection::kInput:
+      return xls_vast_module_port_direction_input;
+    case xls::verilog::ModulePortDirection::kOutput:
+      return xls_vast_module_port_direction_output;
+  }
+  LOG(FATAL) << "Invalid ModulePortDirection encountered.";
+}
+
+xls_vast_def* xls_vast_verilog_module_port_get_def(
+    struct xls_vast_module_port* port) {
+  auto* cpp_port = reinterpret_cast<xls::verilog::ModulePort*>(port);
+  return reinterpret_cast<xls_vast_def*>(cpp_port->wire);
+}
+
+char* xls_vast_def_get_name(struct xls_vast_def* def) {
+  auto* cpp_def = reinterpret_cast<xls::verilog::Def*>(def);
+  return xls::ToOwnedCString(cpp_def->GetName());
+}
+
+xls_vast_data_type* xls_vast_def_get_data_type(struct xls_vast_def* def) {
+  auto* cpp_def = reinterpret_cast<xls::verilog::Def*>(def);
+  return reinterpret_cast<xls_vast_data_type*>(cpp_def->data_type());
+}
+
+bool xls_vast_data_type_width_as_int64(struct xls_vast_data_type* type,
+                                       int64_t* out_width, char** error_out) {
+  CHECK(out_width != nullptr);
```

Comment:
nit: are we sure we want to `CHECK` here vs error'ing out? (it seems to be the first time we do this in this file)

Reply:
I was mostly curious if that was the first case of us doing that :)

If that's already being done in other file, it's consistent to do it here as well!